File: pkg/data/render_test.go
package data

import (
	"fmt"
	"io"
	"strings"
	"testing"
)

type MockWriterRenderer struct {
	w io.Writer
}

func (r *MockWriterRenderer) logf(format string, args ...any) {
	fmt.Fprintf(r.w, format+"\n", args...)
}

func (r *MockWriterRenderer) BeginForm() {
	r.logf("BeginForm")
}
func (r *MockWriterRenderer) EndForm() {
	r.logf("EndForm")
}
func (r *MockWriterRenderer) BeginFieldset(label string) {
	r.logf("BeginFieldset label=%s", label)
}
func (r *MockWriterRenderer) EndFieldset() {
	r.logf("EndFieldset")
}
func (r *MockWriterRenderer) Input(label, name, inputType, value string) {
	r.logf("Input label=%s name=%s type=%s value=%s", label, name, inputType, value)
}
func (r *MockWriterRenderer) Checkbox(label, name string, checked bool) {
	r.logf("Checkbox label=%s name=%s checked=%t", label, name, checked)
}
func (r *MockWriterRenderer) Select(label, name string, options []string, value string) {
	r.logf("Select label=%s name=%s options=%v value=%s", label, name, options, value)
}

func TestRenderDefinition(t *testing.T) {
	tests := map[string]struct {
		Input    any
		Expected []string
	}{

		"nested object": {
			Input: map[string]any{
				"profile": map[string]any{
					"name": map[string]any{
						"first": "Jane",
						"last":  "Doe",
					},
					"age": 42,
				},
			},
			Expected: []string{
				"BeginForm name=form action=/submit method=POST",
				"BeginFieldset label=profile",
				"Input label=age name=profile.age type=number value=42",
				"BeginFieldset label=name",
				"Input label=first name=profile.name.first type=text value=Jane",
				"Input label=last name=profile.name.last type=text value=Doe",
				"EndFieldset",
				"EndFieldset",
				"EndForm Submit",
			},
		},
	}

	for name, tt := range tests {
		t.Run(name, func(t *testing.T) {
			var buf strings.Builder
			mock := &MockWriterRenderer{w: &buf}
			Render(tt.Input, mock)
			lines := strings.Split(strings.TrimSpace(buf.String()), "\n")
			compareCallStacks(t, lines, tt.Expected)
		})
	}
}

func compareCallStacks(t *testing.T, got, want []string) {
	t.Helper()

	if len(got) != len(want) {
		t.Logf("Length mismatch: got %d calls, want %d calls", len(got), len(want))
	}

	max := len(got)
	if len(want) > max {
		max = len(want)
	}

	hasDiff := false

	for i := 0; i < max; i++ {
		var g, w string
		if i < len(got) {
			g = strings.TrimSpace(got[i])
		}
		if i < len(want) {
			w = strings.TrimSpace(want[i])
		}

		status := "✓"
		if g != w {
			status = "✗"
			hasDiff = true
		}

		t.Logf("[%s] [%2d] want: %-60s got: %s", status, i, w, g)
	}

	if hasDiff {
		t.FailNow()
	}
}

File: front/tests/render.test.ts
import { describe, it, expect } from "vitest";
import { render, WriterRenderer, Dict } from "../src/pkg/data/render";

class MockWriterRenderer implements WriterRenderer {
  private lines: string[] = [];

  private log(line: string): void {
    this.lines.push(line);
  }

  beginForm(): void {
    this.log("BeginForm");
  }

  endForm(): void {
    this.log("EndForm");
  }

  beginObject(label: string): void {
    this.log(`BeginFieldset label=${label}`);
  }

  endObject(): void {
    this.log("EndFieldset");
  }

  input(label: string, name: string, inputType: string, value: string): void {
    this.log(
      `Input label=${label} name=${name} type=${inputType} value=${value}`,
    );
  }

  checkbox(label: string, name: string, checked: boolean): void {
    this.log(
      `Checkbox label=${label} name=${name} checked=${checked.toString()}`,
    );
  }

  select(label: string, name: string, options: string[], value: string): void {
    this.log(
      `Select label=${label} name=${name} options=${JSON.stringify(
        options,
      )} value=${value}`,
    );
  }

  getOutput(): string[] {
    return this.lines.slice();
  }
}

function compareStacks(got: string[], want: string[]) {
  expect(got.length).toBe(want.length);
  for (let i = 0; i < want.length; i++) {
    expect(got[i]).toBe(want[i]);
  }
}

describe("render", () => {
  it("nested object", () => {
    const input: Dict = {
      profile: {
        name: {
          first: "Jane",
          last: "Doe",
        },
        age: 42,
      },
    };

    const expectLines = [
      "BeginForm",
      "BeginFieldset label=profile",
      "Input label=age name=profile.age type=number value=42",
      "BeginFieldset label=name",
      "Input label=first name=profile.name.first type=text value=Jane",
      "Input label=last name=profile.name.last type=text value=Doe",
      "EndFieldset",
      "EndFieldset",
      "EndForm",
    ];

    const mock = new MockWriterRenderer();
    const err = render(input, mock);

    expect(err).toBeNull();
    compareStacks(mock.getOutput(), expectLines);
  });

  it("throws on array", () => {
    const input: Dict = {
      profile: [1, 2, 3],
    };

    const mock = new MockWriterRenderer();
    const err = render(input, mock);

    expect(err).not.toBeNull();
  });
});
